src/analysis/syntax_doc.ml

```diff
@@ -0,0 +1,231 @@
+      Some
+        {
+          name = "Destructive substitution";
+          description =
+            "Behaves like normal signature constraints but removes the \
+             redefined type or module from the signature.";
+          documentation =
+            syntax_doc_url
+              "signaturesubstitution.html#ss:destructive-substitution";
+        }
+  | (_, Type_kind _)
+    :: (_, Type_declaration _)
+    :: (_, Signature_item ({ sig_desc = Tsig_typesubst _; _ }, _))
+    :: _ ->
+      Some
+        {
+          name = "Local substitution";
+          description =
+            "Behaves like destructive substitution but is introduced during \
+             the specification of the signature, and will apply to all the \
+             items that follow.";
+          documentation =
+            syntax_doc_url "signaturesubstitution.html#ss:local-substitution";
+        }
+  | (_, Module_type _)
+    :: (_, Module_type _)
+    :: ( _,
+         Module_type_constraint
+           (Tmodtype_explicit
+             { mty_desc = Tmty_with (_, [ (_, _, Twith_modtype _) ]); _ }) )
+    :: _ ->
+      Some
+        {
+          name = "Module substitution";
+          description =
+            "Behaves like type substitutions but are useful to refine an \
+             abstract module type in a signature into a concrete module type,";
+          documentation =
+            syntax_doc_url
+              "signaturesubstitution.html#ss:module-type-substitution";
+        }
+  | (_, Type_kind Ttype_open)
+    :: (_, Type_declaration { typ_private = Public; _ })
+    :: _ ->
+      Some
+        {
+          name = "Extensible variant type";
+          description =
+            "Can be extended with new variant constructors using `+=`.";
+          documentation = syntax_doc_url "extensiblevariants.html";
+        }
+  | (_, Constructor_declaration _)
+    :: (_, Type_kind (Ttype_variant _))
+    :: (_, Type_declaration { typ_private = Public; _ })
+    :: _
+  | _
+    :: (_, Constructor_declaration _)
+    :: (_, Type_kind (Ttype_variant _))
+    :: (_, Type_declaration { typ_private = Public; _ })
+    :: _ ->
+      Some
+        {
+          name = "Variant Type";
+          description =
+            "Represent data that may take on multiple different forms.";
+          documentation = syntax_doc_url "typedecl.html#ss:typedefs";
+        }
+  | (_, Type_kind Ttype_abstract)
+    :: (_, Type_declaration { typ_private = Public; typ_manifest = None; _ })
+    :: _ ->
+      Some
+        {
+          name = "Abstract type";
+          description =
+            "Define variants with arbitrary data structures, including other \
+             variants, records, and functions";
+          documentation = syntax_doc_url "typedecl.html#ss:typedefs";
+        }
+  | (_, Core_type _)
+    :: (_, Core_type _)
+    :: (_, Label_declaration _)
+    :: (_, Type_kind (Ttype_record _))
+    :: (_, Type_declaration { typ_private = Public; _ })
```

Comment:
We have lot of duplication depending on the type being public or private. You might want to refactor it: you could match for both public and private and then you update the texts with privacity information:

```
let name =" Record type" in
let name = match typ_private with
  | Public -> name
  | Private -> Format.sprintf "Private %s" name
in

let description = "Record type description." in
let name = match typ_private with
  | Public -> name
  | Private -> Format.sprintf "%s This type is private, values cannot be constructed directly but can be de-structured as usual." name
in
... 
```

Reply:
This was a fun refactoring to do